GVFS/FastFetch/IndexLock.cs

```diff
@@ -0,0 +1,108 @@
+﻿using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace FastFetch
+{
+    /// <summary>
+    ///   A mechanism for holding the 'index.lock' on a repository for the time it takes to update the index
+    ///   and working tree.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This class should not have to exist.  If FastFetch was in compliance with the git way of doing
+    ///     business, then <see cref="CheckoutPrefetcher"/> would work like this:
+    ///   </para>
+    ///   <list type="bullet">
+    ///     <item>
+    ///       It would open index.lock like this does - with CreateNew, before it started messing with the working tree.
+    ///     </item>
+    ///     <item>
+    ///       It would have just one class responsible for writing the new index into index.lock (now it has two,
+    ///       <see cref="Index"/> and <see cref="GitIndexGenerator"/>).  And this combined class would write in the
+    ///       file size and timestamp information from the appropriate sources as it goes.
+    ///     </item>
+    ///     <item>
+    ///       It would then reread index.lock (without closing it) and calculate the hash.
+    ///     </item>
+    ///     <item>
+    ///       It would then delete the old index file, close index.lock, and move it to index.

```

Comment:
Not so easy to pull off here - first there are 2 classes that are in the index-rewriting business, plus each of those closes the file and re-opens it to hash them.  The first of those index-rewriters (GitIndexGenerator) leaves the index in a bad spot, because any subsequent 'git status' operation will re-hash the whole repo (taking an hour and change).  And then there's the fact that 'head' and 'index' really have to change in lock-step in order for subsequent fastfetch'es to work correctly.  There's no defense against that, right now.